File: Contest/14th May Weekly/Q4.go
package main

import "fmt"

func dfs(course int, visited map[int]bool, coursesMapping map[int][]int) bool {
	if len(coursesMapping[course]) == 0 {
		return true
	}
	if _, present := visited[course]; present {
		return false
	}
	visited[course] = true
	for _, dependancy := range coursesMapping[course] {
		return dfs(dependancy, visited, coursesMapping)
	}

	return true

}

func maxMoves(grid [][]int) int {
	dp := make([][]int, len(grid)+1)
	for i, _ := range dp {
		dp[i] = make([]int, len(grid[0])+1)
	}

	rows := len(grid)
	cols := len(grid[0])

	for i := 0; i < rows; i++ {
		for j := 0; j < cols-1; j++ {
			// if grid[i][j+1] > grid[i][j]{

			// }
			if grid[i][j+1] > grid[i][j] {
				dp[i][j]++
			}
			if i < rows-1 && grid[i+1][j+1] > grid[i][j] {
				dp[i][j]++
			}
			if i > 0 && grid[i-1][j+1] > grid[i][j] {
				dp[i][j]++
			}
			// dp[i][j] = dp[i+1][j+1] + dp[i-1][j+1]
			// if i == rows-1 && j == cols-1 {
			// 	dp[i][j] = 0
			// 	continue
			// }
		}

	}
	fmt.Println(dp)
	return 0
}

func dfsv2(node int, visited map[int]bool, adj map[int][]int) {
	if visited[node] {
		return
	}
	visited[node] = true
	for _, ad := range adj[node] {
		visited[ad] = true
		dfs(ad, visited, adj)
	}
	return

}
func countCompleteComponents(n int, edges [][]int) int {
	adj := make(map[int][]int)
	for i := 0; i < n; i++ {
		adj[i] = []int{}
	}
	for _, edge := range edges {
		adj[edge[0]] = append(adj[edge[0]], edge[1])
		adj[edge[1]] = append(adj[edge[1]], edge[0])
	}
	visited := make(map[int]bool)
	count := 0
	for i, _ := range adj {
		if !visited[i] {
			count++
			dfsv2(i, visited, adj)
		}
	}
	return count
}

func main() {
	// target := 10
	nums := [][]int{{0, 1}, {0, 2}, {1, 2}, {3, 4}}
	n := 6
	fmt.Println(countCompleteComponents(n, nums))
}

File: Contest/28th May Weekly/Q4.py
from typing import List


class Solution:
    def maxIncreasingCells(self, mat: List[List[int]]) -> int:
        ROWS = len(mat)
        COLS = len(mat[0])

        def dfs(r, c, cost):
            # if r == ROWS - 1 or c == COLS - 1 or r == 0 or c == 0:
            #     return 1

            if r > 0 and mat[r - 1][c] > mat[r][c]:
                cost = max(cost, dfs(r - 1, c, cost + 1))
            if c > 0 and mat[r][c - 1] > mat[r][c]:
                cost = max(cost, dfs(r, c - 1, cost + 1))
            if c < COLS - 1 and mat[r][c + 1] > mat[r][c]:
                cost = max(cost, dfs(r, c + 1, cost + 1))
            if r < ROWS - 1 and mat[r + 1][c] > mat[r][c]:
                cost = max(cost, dfs(r + 1, c, cost + 1))

            return cost

        maxVal = 0
        for i in range(ROWS):
            for j in range(COLS):
                maxVal = max(maxVal, dfs(i, j, 1))

        return maxVal


# target =
obj = Solution
nums = [[3, 1, 6], [-9, 5, 7]]

print(obj.maxIncreasingCells(obj, nums))
